File: causal_model/crates/cpcm-factors/src/instruments.rs
use std::collections::HashMap;

/// Compute instrumental variables from the data panel.
///
/// Each IV is lagged by >= 1 day to avoid simultaneity.
/// Returns a map of iv_name → Vec<f64> and a map of factor → iv_name.
pub fn compute_instruments(
    data: &HashMap<String, Vec<f64>>,
    n_dates: usize,
) -> (HashMap<String, Vec<f64>>, HashMap<String, String>) {
    let mut iv_data = HashMap::new();
    let mut iv_map = HashMap::new(); // factor -> iv_name

    // gas_spike → instruments for liq_flow
    // Z-score of gas diff > 2σ, lagged by 1 day
    // Try multiple column names: Dune query 6831658 returns avg_base_fee_gwei, not avg_gas_price_gwei
    let gas_col = data.get("eth_avg_gas_price_gwei")
        .or_else(|| data.get("eth_avg_base_fee_gwei"))
        .or_else(|| data.get("eth_FeeTotNtv"));
    if let Some(gas) = gas_col {
        let spike = lagged_z_score_spike(gas, 2.0, 1);
        iv_data.insert("gas_spike".to_string(), spike);
        iv_map.insert("liq_flow".to_string(), "gas_spike".to_string());
    }

    // liquidation_level → instruments for funding_basis
    // Liquidation volume spikes, lagged by 1 day
    if let Some(liq) = data.get("eth_liquidation_volume_usd") {
        let spike = lagged_z_score_spike(liq, 2.0, 1);
        iv_data.insert("liquidation_level".to_string(), spike);
        iv_map.insert("funding_basis".to_string(), "liquidation_level".to_string());
    }

    // stablecoin_mint → instruments for stable_flow
    // Lagged daily change in stablecoin supply
    let supply_cols = ["usdc_SplyCur", "usdt_SplyCur"];
    let available: Vec<&Vec<f64>> = supply_cols
        .iter()
        .filter_map(|&col| data.get(col))
        .collect();

    if !available.is_empty() {
        let total: Vec<f64> = (0..n_dates)
            .map(|i| {
                available
                    .iter()
                    .map(|col| if i < col.len() && !col[i].is_nan() { col[i] } else { 0.0 })
                    .sum()
            })
            .collect();
        let mint = lag(&diff(&total), 1);
        iv_data.insert("stablecoin_mint".to_string(), mint);
        iv_map.insert("stable_flow".to_string(), "stablecoin_mint".to_string());
    }

    // protocol_event → instruments for chain_congestion
    // Fee regime change: abs(diff(fees)) > 2σ, lagged by 1 day
    let fees_col = data.get("eth_fees_usd")
        .or_else(|| data.get("eth_FeeTotNtv"));
    if let Some(fees) = fees_col {
        let spike = lagged_z_score_spike(fees, 2.0, 1);
        iv_data.insert("protocol_event".to_string(), spike);
        iv_map.insert("chain_congestion".to_string(), "protocol_event".to_string());
    }

    (iv_data, iv_map)
}

/// Compute a z-scored spike indicator: 1.0 if |z(diff(x))| > threshold, else 0.0.
/// Then lag by `lag_days`.
fn lagged_z_score_spike(x: &[f64], threshold: f64, lag_days: usize) -> Vec<f64> {
    let d = diff(x);
    let z = z_score_abs(&d);
    let indicator: Vec<f64> = z
        .iter()
        .map(|&v| {
            if v.is_nan() {
                f64::NAN
            } else if v > threshold {
                1.0
            } else {
                0.0
            }
        })
        .collect();
    lag(&indicator, lag_days)
}

/// First-difference.
fn diff(x: &[f64]) -> Vec<f64> {
    if x.is_empty() {
        return vec![];
    }
    let mut d = vec![f64::NAN];
    for i in 1..x.len() {
        d.push(if x[i].is_nan() || x[i - 1].is_nan() {
            f64::NAN
        } else {
            x[i] - x[i - 1]
        });
    }
    d
}

/// Z-score of absolute values: |x_i - mean| / std.
fn z_score_abs(x: &[f64]) -> Vec<f64> {
    let valid: Vec<f64> = x.iter().filter(|v| !v.is_nan()).copied().collect();
    if valid.is_empty() {
        return x.to_vec();
    }
    let abs_vals: Vec<f64> = valid.iter().map(|v| v.abs()).collect();
    let mean = abs_vals.iter().sum::<f64>() / abs_vals.len() as f64;
    let var = abs_vals
        .iter()
        .map(|v| (v - mean).powi(2))
        .sum::<f64>()
        / abs_vals.len() as f64;
    let std = var.sqrt();
    if std < 1e-15 {
        return vec![0.0; x.len()];
    }
    x.iter()
        .map(|&v| {
            if v.is_nan() {
                f64::NAN
            } else {
                (v.abs() - mean) / std
            }
        })
        .collect()
}

/// Lag a series by `k` positions (shift forward). First `k` elements become NaN.
fn lag(x: &[f64], k: usize) -> Vec<f64> {
    let mut result = vec![f64::NAN; x.len()];
    for i in k..x.len() {
        result[i] = x[i - k];
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lag() {
        let x = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let lagged = lag(&x, 1);
        assert!(lagged[0].is_nan());
        assert!((lagged[1] - 1.0).abs() < 1e-10);
        assert!((lagged[4] - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_diff() {
        let x = vec![10.0, 12.0, 11.0];
        let d = diff(&x);
        assert!(d[0].is_nan());
        assert!((d[1] - 2.0).abs() < 1e-10);
        assert!((d[2] - (-1.0)).abs() < 1e-10);
    }

    #[test]
    fn test_compute_instruments_returns_map() {
        let mut data = HashMap::new();
        data.insert(
            "eth_avg_gas_price_gwei".to_string(),
            vec![10.0, 12.0, 50.0, 11.0, 13.0, 100.0, 10.0, 12.0, 11.0, 10.0],
        );

        let (iv_data, iv_map) = compute_instruments(&data, 10);
        assert!(iv_data.contains_key("gas_spike"));
        assert_eq!(iv_map.get("liq_flow"), Some(&"gas_spike".to_string()));
    }
}

File: causal_portfolio/factors/instruments.py
"""Instrumental-variable construction for 2SLS — Python port of instruments.rs.

The DAG declares four instruments, each lagged >= 1 day to break simultaneity:

    gas_spike          instruments  liq_flow
    liquidation_level  instruments  funding_basis
    stablecoin_mint    instruments  stable_flow
    protocol_event     instruments  chain_congestion

Each is built from warehouse columns the same way the Rust reference does:
a lagged spike indicator (ROLLING z-score of the day-over-day change > 2σ,
window 252 / min_periods 60, so the label at t uses only data ≤ t), except
stablecoin_mint which is the lagged level of net minting (diff of supply).

Column names here are mapped to what our warehouse actually has (the Rust code
hard-coded ETH Dune column names that differ slightly). Where a single ETH
series is unavailable we aggregate across assets.

IMPORTANT: these are deliberately weak-by-nature signals (mostly-zero spike
indicators). 2SLS is only valid if the first-stage F-statistic shows the
instrument actually moves its treatment factor — that gate lives in the 2SLS
diagnostics, not here.
"""

from __future__ import annotations

import numpy as np
import pandas as pd

from causal_portfolio.factors.builder import FACTOR_SOURCE_ASSETS

# treatment_factor -> (instrument_name, builder_kind, candidate_source_columns)
# builder_kind: "spike" = lagged z-score spike indicator; "level_diff" = lagged diff
_INSTRUMENT_SPEC = {
    "liq_flow": ("gas_spike", "spike",
                 ["avg_gas_utilization", "stddev_base_fee_gwei", "avg_base_fee_gwei", "FeeTotNtv"]),
    "funding_basis": ("liquidation_level", "spike",
                      ["liquidation_volume_usd", "perp_liquidation", "avg_liquidation_usd"]),
    "stable_flow": ("stablecoin_mint", "level_diff",
                    ["SplyCur", "stablecoin_circulating_usd"]),
    "chain_congestion": ("protocol_event", "spike",
                         ["fees", "total_fees_usd", "base_fees", "FeeTotNtv"]),
}

# Stablecoin assets whose supply we sum for stablecoin_mint
_STABLE_ASSETS = FACTOR_SOURCE_ASSETS


# ── helpers (faithful to instruments.rs) ────────────────────────────


def _diff(x: np.ndarray) -> np.ndarray:
    """First difference; first element NaN, NaN-propagating."""
    d = np.full(len(x), np.nan)
    d[1:] = x[1:] - x[:-1]
    return d


def _z_score_abs(x: np.ndarray, window: int = 252, min_periods: int = 60) -> np.ndarray:
    """ROLLING z-score of absolute values: (|x_t| - mean_t) / std_t over a
    trailing window (population std, ddof=0), ignoring NaN.

    Trailing window ⇒ the label at t uses only data ≤ t (the previous
    full-sample version leaked future data into spike thresholds). NaN until
    `min_periods` non-NaN values are in the window. Matches the Rust
    rolling_z_score_abs in cpcm-factors/instruments.rs.
    """
    s = pd.Series(np.abs(np.asarray(x, dtype=float)))
    mean = s.rolling(window, min_periods=min_periods).mean()
    std = s.rolling(window, min_periods=min_periods).std(ddof=0)
    z = (s - mean) / std
    z[(std < 1e-15) & s.notna() & mean.notna()] = 0.0
    return z.to_numpy()


def _lag(x: np.ndarray, k: int) -> np.ndarray:
    """Shift forward by k; first k elements NaN."""
    out = np.full(len(x), np.nan)
    if k < len(x):
        out[k:] = x[:-k] if k > 0 else x
    return out


def _lagged_z_score_spike(
    x: np.ndarray, threshold: float = 2.0, lag: int = 1,
    window: int = 252, min_periods: int = 60,
) -> np.ndarray:
    """1.0 if rolling-z(diff(x)) > threshold else 0.0, then lagged by `lag`."""
    z = _z_score_abs(_diff(x), window=window, min_periods=min_periods)
    ind = np.where(np.isnan(z), np.nan, (z > threshold).astype(float))
    return _lag(ind, lag)


def _find_columns(panel: pd.DataFrame, suffixes: list[str]) -> list[str]:
    """All panel columns ending in any of the given metric suffixes."""
    cols = []
    for suf in suffixes:
        cols.extend([c for c in panel.columns if c.endswith(f"_{suf}") or c == suf])
        if cols:
            break  # use the first available source family, like the Rust fallback chain
    return cols


# ── public API ──────────────────────────────────────────────────────


def build_instruments(panel: pd.DataFrame) -> tuple[pd.DataFrame, dict[str, str]]:
    """Build instrument time series from a wide panel.

    Returns:
        (instruments_df, iv_map) where instruments_df has one column per
        successfully-built instrument (indexed like panel), and iv_map maps
        treatment_factor -> instrument_name for the factors we could instrument.
    """
    out: dict[str, np.ndarray] = {}
    iv_map: dict[str, str] = {}

    for factor, (iv_name, kind, sources) in _INSTRUMENT_SPEC.items():
        if factor == "stable_flow":
            # stablecoin_mint = lagged diff of summed stablecoin supply
            supply_cols = [c for c in panel.columns
                           if any(c.startswith(f"{a}_") for a in _STABLE_ASSETS)
                           and (c.endswith("_SplyCur") or c.endswith("_stablecoin_circulating_usd"))]
            if not supply_cols:
                continue
            total = panel[supply_cols].sum(axis=1).values.astype(float)
            out[iv_name] = _lag(_diff(total), 1)
            iv_map[factor] = iv_name
            continue

        cols = _find_columns(panel, sources)
        if not cols:
            continue
        # Aggregate the source across available assets (sum), then build spike
        series = panel[cols].sum(axis=1).values.astype(float)
        out[iv_name] = _lagged_z_score_spike(series, threshold=2.0, lag=1)
        iv_map[factor] = iv_name

    instruments_df = pd.DataFrame(out, index=panel.index)
    return instruments_df, iv_map
